README.md
# webdriver

A simple (features as needs arise) Clojure wrapper of Selenium Webdriver

## Usage
[![Clojars Project](https://img.shields.io/clojars/v/webdriver.svg)](https://clojars.org/webdriver)
[![CircleCI](https://circleci.com/gh/komcrad/webdriver/tree/master.svg?style=svg&circle-token=a5fcd5b0389dd482ec5e55fb3c6bab0715377cd9)](https://circleci.com/gh/komcrad/webdriver/tree/master)

If you want to use the truly headless/recording environment, you'll need to be running a linux host with xvfb and ffmpeg installed.  
`sudo apt-get install ffmpeg xvfb`  
`sudo pacman -S ffmpeg xorg-server-xvfb`

## Example
### Assuming you're in a repl (`lein repl`)...

Import the core namespace:

`(use 'webdriver.core)`

Create a driver object

Chrome:
- `(def driver (create-driver {:driver-type :chrome}))`

Firefox:
- `(def driver (create-driver {:driver-type :firefox}))`

Specific driver versions can also be passed to :chrome or :firefox versions
- `(def driver (create-driver {:driver-type :chrome :version "81.0.4044.138"}))`

Headless:
- `(def driver (create-driver {:driver-type :chrome :driver-args ["--headless"]}))`

Xvfb (Linux only with xvfb installed)

```
(require '[webdriver.screen :as scr])
(def screen (scr/start-screen)) ;this is separate from driver so that we can have multiple drivers run in a single frame buffer
(def driver (create-driver {:driver-type :chrome :xvfb-port (:xvfb-port screen)}))
```

Now you can pass the driver object into the other functions in core to manipulate it.

```
(to driver "https://google.com")
(set-element driver :name "q" "silly memes")
(click (wait-for-element driver :xpath "//input[@value = 'Google Search'][1]"))
```

From functional code:
```
(with-webdriver [driver :driver-type :chrome :driver-args ["--headless"]]
  (to driver "https://google.com")
  (set-element driver :name "q" "silly memes")
  (click (wait-for-element driver :xpath "//input[@value = 'Google Search'][1]")))
```

If you want a truly headless environment that doesn't have the issues of geckodriver --headless or chromedriver --headless:
```
(scr/with-screen [screen]
  (with-webdriver [driver :driver-type :chrome :xvfb-port (:xvfb-port screen)]
    (to driver "https://google.com")
    (set-element driver :name "q" "silly memes")
    (click (wait-for-element driver :xpath "//input[@value = 'Google Search'][1]"))))
```

If you want to record that headless session:
```
(scr/with-recorded-screen [screen :vid-out "/tmp/webdriver.mp4"]
  (with-webdriver [driver :driver-type :chrome :xvfb-port (:xvfb-port screen)]
    (to driver "https://google.com")
    (set-element driver :name "q" "silly memes")
    (click (wait-for-element driver :xpath "//input[@value = 'Google Search'][1]"))))
```

etc...

core.clj contains functions to handle common browser tasks. You can either read through that or checkout tests/webdriver/core_test.clj for unit test examples.

## Breaking changes:
  - In webdriver 0.11.0, functions that use to return webdrivers now return maps with a WebDriver object at :driver. This means that some code that used java interop on driver objects will now be broken. Internally this has been updated. According to the unit tests, the functions in webdriver.core still work together properly.
  ```
  ; in webdriver 0.10.0 this would work
  (def driver (create-driver {:driver-type :chrome}))
  (.getVersion (.getCapabilities driver))
  (driver-quit driver)

  ; in webdriver 0.11.0 you'll have to do it differently
  (def driver (create-driver {:driver-type :chrome}))
  (.getVersion (.getCapabilities (:driver driver))) ; notice (:driver driver) instead of just driver
  (driver-quit driver)
  ```

## Building Docker image
```
docker build --no-cache -t attackhelicopter/webdriver:0.11.0 -t attackhelicopter/webdriver:latest .
```

## Running unit tests in docker image locally
```
docker run -v /dev/shm/:/dev/shm -it attackhelicopter/webdriver:latest /bin/bash
git clone https://github.com/komcrad/webdriver.git
cd webdriver
lein test
```

## Tested versions of firefox and chrome
- webdriver 0.17.0
  - Firefox 76.0.1
  - Chromium 83.0.4103.106
- webdriver 0.15.0
  - Firefox 70.0.1
  - Chromium 78.0.3904.97
- webdriver 0.14.0
  - Firefox 68.0.1
  - Chromium 76.0.3809.87
- webdriver 0.13.0
  - Firefox 67.0.1
  - Chromium 75.0.3770.80
- webdriver 0.12.0
  - Firefox 66.0.1
  - Google Chrome 74.0.3729.108
- webdriver 0.11.0
  - Firefox 66.0
  - Google Chrome 73.0.3683.86
- webdriver 0.10.0
  - Firefox 66.0
  - Firefox 65.0.1
  - Google Chrome 73.0.3683.86
  - Google Chrome 73.0.3683.75
- webdriver 0.8.1, 0.9.0, 0.10.0
  - Firefox 65.0
  - Google Chrome 72.0.3626.81
- webdriver 0.7.0
  - Firefox 63.0.3
  - Google Chrome 71.0.3578.80
- webdriver 0.6.0
  - Firefox: 62.0.3
  - Google Chrome 70.0.3538.67
- webdriver 0.5.3
  - Firefox: 62.0
  - Google Chrome 69.0.3497.100
- webdriver 0.5.2
  - Firefox: 61.0.1
  - Google Chrome 68.0.3440.106


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [0.17.0 2020-06-16]
## Added
- arch linux build for ci
- Upgraded default chromedriver to 83.0.4103.106
- Upgraded to clojure 1.10.1

## [0.14.1 2019-08-14]
## Fixed
- added --no-sandbox to chrome options which fixes issues with running chromium in docker

## [0.14.0 2019-08-05]
## Added
- default driver is 76.0.3809.68

## Fixed
- screen-shot now accepts the driver map that (create-driver produces)

## [0.12.1]
## Fixed
- download-driver function will now extract the tar.gz file instead of trying to unzip it

## [0.12.0 2019-04-28]
## Added
- Ability to deal with webdriver screen sizes

## [0.11.0 2019-03-28]
## Added
- Truly headless support for linux. Allows firefox and chrome to run in xvfb. Everything is handled internally.
- Ability to record drivers that run in xvfb.

## [0.10.0 2019-02-15]
## Added
- better file download support
- selected? function

## [0.9.0 2019-02-04]
## Added
- enabled? function so people don't have to use java interop or the is-visible function to see if an element is enabled

## [0.8.1 2019-01-31]
## Updated
- browser version used for circle ci
- driver versions so newer browsers will be supported

## [0.8.0 2019-01-30]
## Added
- sibling and parent functions for traversing the dom
- select-elm functions for dealing with select elements

## [0.7.0 - 2019-01-10]
### Added
- insert-html and delete-elm functions for basic dom manipulation
- html function: a wrapper around the html macro in the hiccup library
- wait-for-trans function to allow waiting on transitions to finish
- options function for getting available options of a select element

## [0.6.2 - 2018-11-21]
### Fixed/Added
- Added to-localhost function that allows firefox to local localhost without throwing an exception

## [0.6.1 - 2018-11-14]
### Fixed
- Firefox is now able to set-file-input

## [0.6.0 - 2018-10-26]
### Added
- with-webdriver: a macro to replace with-driver with better syntax
- poll-interval option for wait-for-element
- visible? function
### Changed
- Updated version of chromedriver to 2.43
- Updated geckdriver to 0.23.0
- Updated circleci to use latest webdriver docker image containing chrome 70 and firefox 62
### Fixed
- is-visible docstring

## [0.5.3 - 2018-09-21]
### Changed
- Updates version of chromedriver to 2.42
- Update circleci to use new webdriver docker image that contains updated web browsers

## [0.5.2 - 2018-09-04]
### Fixed
- Specifies version of chromedriver that will be used

## [0.5.1 - 2018-08-21]
### Fixed
- Issue where --headless chromedriver would not ignore insecure certs
- missing function wait-elm-dom. A function to wait for elment to attach to dom without needing to be clickable.

## [0.5.0 - 2018-08-03]
### Added
- New set-elms function that gives the option to use a different syntax for setting multiple elements
## [0.4.0 - 2018-06-25]
### Added
- functions for dealing with invisible and unclickable elements (wait-q and try-click)

## [0.3.2 - 2018-05-29]
### Fixed
- annoying chromedriver startup logging

## [0.3.1 - 2018-05-04]
### Fixed
- an issue with scroll-into-view not scrolling to center of view

## [0.3.0 - 2018-05-03]
### Added
- added scroll-into-view in all functions where element being visible is important
- scroll-into-view function
- css function
- attr function
- wait-click function
- ability to click multiple elements by passing in multiple element selectors to core/click

## [0.2.5 - 2018-02-05]
### Added
- function to set file input elements
- wrapper around sendKeys
- function for taking screenshots of the driver

## [0.2.4 - 2018-01-24]
### Added
- Waits for alerts to exist for up to 2 seconds and waits 10ms for the alert to close
- Downloads go to current directory (for firefox there is a whitelist of MIME types that will download without prompt)
- Chrome will always ignore cert errors

## [0.2.3 - 2018-01-23]
### Changed
- Changes how elements are set so they will trigger key-up events

## [0.2.2 - 2018-01-19]
### Added
- set-elements option to query for elements

## [0.2.1 - 2018-01-17]
### Added
- get-visible-element (function that uses get-elements and returns the first visible element)

### Changed
- reverted 0.2.0 changes
- click, set-element, and clear all use get-visible-element for queries


## [0.2.0 - 2018-01-17]
### Changed
- the get-element function will now only return an element if it is visible.. side-effects include only being able to call (click driver :name "button") (set driver :name "input" "value") on visible elements. You can still pass in webelements. To functions that use get-element

## [0.1.1] - 2017-12-06
### Added
- Added cookie support

## 0.1.0 - 2017-12-05
### Added
- Released a basic clojure wrapper for the selenium library

[Unreleased]: https://github.com/komcrad/webdriver/compare/0.1.1...HEAD
[0.2.0]: https://github.com/your-name/webdriver/compare/0.1.1...0.2.0
[0.1.1]: https://github.com/komcrad/webdriver/compare/0.1.0...0.1.1


test/glados_webdriver/core_test.clj
(ns glados-webdriver.core-test
  (:require [clojure.test :refer :all]
            [glados-webdriver.core :refer :all]
            [clojure.java.io :as io]))
(def test-html-file-url (str "file://" (.getCanonicalPath (io/file "test/resources/index.html"))))

(deftest ^:parallel create-driver-test
  (testing "create-driver"
    (let [driver (create-driver :chrome ["--headless"])]
      (is (= "class org.openqa.selenium.chrome.ChromeDriver" (.toString (type driver))))
      (to driver (str "file://" (.getCanonicalPath (io/file "test/resources/index.html"))))
      (driver-quit driver))
    (let [driver (create-driver :chrome ["--headless"])]
      (is (= "class org.openqa.selenium.chrome.ChromeDriver" (.toString (type driver))))
      (driver-quit driver))
    (let [driver (create-driver :firefox ["--headless"])]
      (is (= "class org.openqa.selenium.firefox.FirefoxDriver" (.toString (type driver))))
      (driver-quit driver))
    (let [driver (create-driver :firefox ["--headless"])]
      (is (= "class org.openqa.selenium.firefox.FirefoxDriver" (.toString (type driver))))
      (driver-quit driver))))

(deftest ^:parallel to-test
  (testing "to"
    (with-all-drivers
      ["--headless"]
      (to driver "https://google.com")
      (is (= "Google Search" (get-element-value driver :name "btnK" :value))))))

(deftest ^:parallel driver-quit-test
  (testing "driver-quit"
    (let [driver (create-driver :chrome)]
      (driver-quit driver)
      (is (thrown? org.openqa.selenium.NoSuchSessionException (to driver "https://google.com"))))
    (let [driver (create-driver :firefox)]
      (driver-quit driver)
      (is (thrown? org.openqa.selenium.NoSuchSessionException (to driver "https://google.com"))))))

(deftest ^:parallel get-elements-test
  (testing "get-elements"
    (with-all-drivers
      ["--headless"]
      (to driver test-html-file-url)
      (is (= 1 (count (get-elements driver :name "p1"))))
      (is (= 2 (count (get-elements driver :id "p2"))))
      (is (= 3 (count (get-elements driver :linkText "paragraph 3"))))
      (is (= 4 (count (get-elements driver :className "p4"))))
      (is (= 5 (count (get-elements driver :xpath "//p[@xpath='p5']"))))
      (is (= 6 (count (get-elements driver :text "paragraph 6"))))
      (is (= 7 (count (get-elements driver :tagName "paragraph")))))))

(deftest ^:parallel get-element-test
  (testing "get-element"
    (with-all-drivers
      ["--headless"]
      (to driver test-html-file-url)
      (is (= "paragraph 1" (.getText (get-element driver :name "p1")))))))

(deftest ^:parallel focused-element-test
  (testing "focused-element")
  (with-all-drivers
    ["--headless"]
    (to driver test-html-file-url)
    (click driver :id "btn1")
    (is (= "Button 1" (get-element-value (focused-element driver) :text)))))

(deftest ^:parallel unfocus-test
    (testing "unfocus")
      (with-all-drivers
        ["--headless"]
        (to driver test-html-file-url)
        (click driver :id "btn1")
        (unfocus driver)
        (is (not (= "Button 1" (get-element-value (focused-element driver) :text))))))

(deftest ^:parallel clear-test
  (testing "clear"
    (with-all-drivers
      ["--headless"]
      (to driver test-html-file-url)
      (set-element driver :id "input1" "somestring")
      (clear driver (get-element driver :id "input1"))
      (is (= "" (get-element-value driver :id "input1" :value)))
      (set-element driver :id "input1" "somestring")
      (clear driver :id "input1")
      (is (= "" (get-element-value driver :id "input1" :value))))))

(deftest ^:parallel implicit-wait-test
  (testing "implicit-wait")
    (with-all-drivers
      ["--headless"]
      (to driver test-html-file-url)
      (click driver :id "btn2")
      (is (thrown? NullPointerException (get-element-value driver :id "input2" :value)))
      (to driver test-html-file-url)
      (implicit-wait driver 10)
      (click driver :id "btn2")
      (is (= "potato" (get-element-value driver :id "input2" :value)))))

(deftest ^:parallel wait-for-element-test
  (testing "wait-for-element"
    (with-all-drivers
      ["--headless"]
      (to driver test-html-file-url)
      (click driver :id "btn2")
      (is (= "potato") (get-element-value (wait-for-element driver :id "input2" 5) :value))
      (click driver :id "btn3")
      (click driver :id "btn2")
      (is (= "potato") (get-element-value (wait-for-element driver :id "input2") :value)))))

(deftest ^:parallel is-visible-test
  (testing "is-visible"
    (with-all-drivers
      ["--headless"]
      (to driver test-html-file-url)
      (is (is-visible driver :id "p2"))
      (is (not (is-visible driver :id "p4315151")))
      (is (is-visible (get-element driver :id "p2")))
      (is (not (is-visible (get-element driver :id "fake news")))))))

(deftest ^:parallel input-text-test
  (testing "input-text"
    (with-all-drivers
      ["--headless"]
      (to driver test-html-file-url)
      (input-text driver (get-element driver :id "input1") "hello there" true)
      (is (= "hello there" (get-element-value driver :id "input1" :value)))
      (input-text driver (get-element driver :id "input1") "world" false)
      (is (= "hello thereworld" (get-element-value driver :id "input1" :value))))))

(deftest ^:parallel set-element-test
  (testing "set-element"
    (with-all-drivers
      ["--headless"]
      (to driver test-html-file-url)
      (set-element driver :id "input1" "hello there")
      (is (= "hello there" (get-element-value driver :id "input1" :value)))
      (set-element driver (get-element driver :id "input1") "hello world")
      (is (= "hello world" (get-element-value driver :id "input1" :value)))
      (set-element driver :id "select1" "Option 2")
      (is (= "Option 2" (get-element-value driver :id "select1" :value))))))

(deftest ^:parallel set-elements-test
  (testing "set-elements"
    (with-all-drivers
      ["--headless"]
      (to driver test-html-file-url)
      (set-elements
        driver
        [(get-element driver :id "input1") (get-element driver :id "select1")]
        ["hello there world" "Option 2"])
      (is (= "hello there world" (get-element-value driver :id "input1" :value)))
      (is (= "Option 2" (get-element-value driver :id "select1" :value))))))

(deftest ^:parallel get-element-value-test
  (testing "get-element-value"
    (with-all-drivers
      ["--headless"]
      (to driver test-html-file-url)
      (set-element driver :id "input1" "hello")
      (set-element driver :id "select1" "Option 3")
      (is (= "hello" (get-element-value driver :id "input1" :value)))
      (is (= "Option 3" (get-element-value driver :id "select1" :value))))))

(deftest ^:parallel click-test
  (testing "click"
    (with-all-drivers
      ["--headless"]
      (to driver test-html-file-url)
      (click driver :id "btn4")
      (is (= "toothpick" (get-element-value driver :id "input6" :value)))
      (to driver "https://google.com")
      (to driver test-html-file-url)
      (click (get-element driver :id "btn4"))
      (is (= "toothpick" (get-element-value driver :id "input6" :value))))))

(deftest ^:parallel alert-text-test
  (testing "alert-text"
    (with-all-drivers []
      (execute-script driver "alert('hello world')")
      (is (= "hello world" (alert-text driver))))))

(deftest ^:parallel alert-accept-test
  (testing "alert-accept"
    (with-all-drivers []
      (to driver test-html-file-url)
      (execute-script driver "alert(alert('hello world'))")
      (is (= "hello world" (alert-text driver)))
      (is (thrown? Exception (click driver :id "btn1")))
      (alert-accept driver)
      (is (= "undefined" (alert-text driver)))
      (alert-accept driver)
      (is (= nil (click driver :id "btn1"))))))

(deftest ^:parallel alert-dismiss-test
  (testing "alert-dismiss"
    (with-all-drivers []
      (to driver test-html-file-url)
      (execute-script driver "alert(confirm('do you like jolly bears?'))")
      (is (= "do you like jolly bears?" (alert-text driver)))
      (alert-accept driver)
      (is (= "true" (alert-text driver)))
      (alert-accept driver)
      (execute-script driver "alert(confirm('do you like jolly bears?'))")
      (is (= "do you like jolly bears?" (alert-text driver)))
      (alert-dismiss driver)
      (is (= "false" (alert-text driver)))
      (alert-accept driver))))

(deftest ^:parallel iframe-test
  (testing "iframe"
    (with-all-drivers ["--headless"]
      (to driver test-html-file-url)
      (is (thrown? java.lang.NullPointerException (get-element-value driver :id "iframeBChild" :text)))
      (iframe driver 1)
      (is (= "I'm child B" (get-element-value driver :id "iframeBChild" :text)))
      (to driver test-html-file-url)
      (iframe driver :id "iframeA")
      (is (= "I'm child A" (get-element-value driver :id "iframeAChild" :text))))))

(deftest ^:parallel iframe-parent-test
  (testing "iframe"
    (with-all-drivers ["--headless"]
      (to driver test-html-file-url)
      (iframe driver :id "iframeB")
      (iframe driver :id "iframeC")
      (is (= "I'm child C" (get-element-value driver :id "iframeCChild" :text)))
      (iframe-parent driver)
      (is (= "I'm child B" (get-element-value driver :id "iframeBChild" :text))))))

(deftest ^:parallel iframe-default-test
  (testing "iframe-default"
    (with-all-drivers ["--headless"]
      (to driver test-html-file-url)
      (iframe driver :id "iframeB")
      (iframe driver :id "iframeC")
      (is (= "I'm child C" (get-element-value driver :id "iframeCChild" :text)))
      (iframe-default driver)
      (is (= "paragraph 1" (get-element-value driver :name "p1" :text))))))


test/webdriver/driver_manager_test.clj
(ns webdriver.driver-manager-test
  (:require [clojure.test :refer :all]
            [webdriver.driver-manager :refer :all]
            [clojure.java.io :as io]
            [clojure.string :as s]
            [digest :as digest]
            [komcrad-utils.io :as kio]))
(deftest download-driver-test
  (testing "download-driver"
    (.delete (download-driver {:driver-type :chrome}))
    (is (.exists (download-driver {:driver-type :chrome})))
    (.delete (download-driver {:driver-type :firefox}))
    (is (.exists (download-driver {:driver-type :firefox})))))


setup.sh
#!/bin/bash
DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
cd $DIR

install-software() {
  software="vim bash-completion git i3-wm i3status i3lock feh rxvt-unicode jdk8-openjdk dmenu brave xclip chromium htop base-devel xdotool"
  sudo pacman -Syu --noconfirm $software
  sudo cp clipboard /usr/lib/urxvt/perl/
}

place-configs () {
  cp -r .bashrc .bg.darkArch.png .vimrc .i3/ .Xresources .Xdefaults ~
}

setup-clojure () {
  curl https://raw.githubusercontent.com/technomancy/leiningen/stable/bin/lein > lein
  chmod +x lein
  sudo mv lein /usr/local/bin/
  yes | sudo lein downgrade 2.9.1
  lein -version
  cp profiles.clj ~/.lein/
}

vim-plugins () {
  mkdir -p ~/.vim/autoload ~/.vim/bundle
  curl -LSso ~/.vim/autoload/pathogen.vim https://tpo.pe/pathogen.vim
  cd ~/.vim/bundle
  git clone https://github.com/tpope/vim-fireplace.git
  git clone https://github.com/ervandew/supertab.git
  cd $DIR
}

touch-pad () {
sudo mkdir -p /etc/X11/xorg.conf.d && sudo tee <<'EOF' /etc/X11/xorg.conf.d/90-touchpad.conf 1> /dev/null
Section "InputClass"
	Identifier "touchpad"
	MatchIsTouchpad "on"
	Driver "libinput"
	Option "Tapping" "on"
    Option "NaturalScrolling" "on"
    Option "ScrollMethod" "twofinger"
EndSection

EOF
}

main () {
  install-software
  place-configs
  setup-clojure
  vim-plugins
  touch-pad
}

main


project.clj
(defproject webdriver "0.17.0"
  :description "A clojure selenium webdriver wrapper"
  :url "https://github.com/komcrad/webdriver"
  :license {:name "LGPL-3.0"
            :url "https://www.gnu.org/licenses/lgpl-3.0.en.html"}
  :dependencies [[org.clojure/clojure "1.10.1"]
                 [org.seleniumhq.selenium/selenium-java "3.141.59"]
                 [io.github.bonigarcia/webdrivermanager "3.3.0"]
                 [org.slf4j/slf4j-simple "1.7.25"]
                 [me.raynes/fs "1.4.6"]
                 [komcrad-utils "0.9.0"]
                 [org.clojure/data.json "0.2.6"]
                 [hiccup "1.0.5"]
                 [http-kit "2.3.0"]
                 [clj-file-zip "0.1.0"]
                 [me.raynes/conch "0.8.0"]]
  :profiles {:dev
             {:dependencies [[digest "1.4.8"]]}})


src/webdriver/screen.clj
(ns webdriver.screen
  (:gen-class)
  (:require [org.httpkit.client :as http]
            [clojure.java.io :as io]
            [clojure.string :as s]
            [komcrad-utils.io :as kio]
            [komcrad-utils.wait :refer [wait-for]]
            [clj-file-zip.core :as zip]
            [me.raynes.conch.low-level :as sh]
            [me.raynes.conch :refer [programs with-programs let-programs]])
  (:import (org.openqa.selenium.logging
            LoggingPreferences LogType)
           (java.util.logging Level Logger)))

(defn screen-port-available? [n]
  (with-programs [ps]
    (empty?
     (filter #(s/includes? % (str "Xvfb :" n))
             (filter #(s/includes? % "Xvfb")
                     (s/split (ps "aux") #"\n"))))))

(defn screen-port []
  (let [port (+ 50 (rand-int 999999950))]
    (if (screen-port-available? port)
      port
      (screen-port))))

(defn start-screen []
  (let [port (screen-port)
        xvfb (sh/proc "Xvfb" (str ":" port) "-screen" "0" "1920x1080x24")]
    {:xvfb xvfb
     :xvfb-port port}))

(defn stop-screen [m]
  (sh/destroy (:xvfb m)))

(defn start-recorder [m]
  {:xvfb-recorder
   (sh/proc "ffmpeg" "-y" "-f" "x11grab" "-video_size" "1920x1080" "-i"
            (str ":" (:xvfb-port m)) "-codec:v" "libx264" "-r"
            "60" (:vid-out m))})

(defn stop-recorder [m]
  (spit (:in (:xvfb-recorder m)) "q")
  (wait-for #(try (.exitValue (:process (:xvfb-recorder m)))
                  (catch Exception e nil)) 10000 10)
  (sh/destroy (:xvfb-recorder m)))

(defmacro with-screen [[screen] & body]
  `(let [~screen (webdriver.screen/start-screen)]
     (try
       ~@body
       (catch Exception e# (throw e#))
       (finally (webdriver.screen/stop-screen ~screen)))))

(defn stream-to-null [is]
  (future
    (try
      (while true
        (.read is) nil)
      (catch Exception e nil))))

(defmacro with-recorded-screen
  [[screen & {:keys [vid-out]
              :as params
              :or {vid-out "/tmp/webdriver.mp4"}}] & body]
  `(let [~screen (webdriver.screen/start-screen)
         recorder# (webdriver.screen/start-recorder
                    (merge ~screen {:vid-out ~vid-out}))]
     (webdriver.screen/stream-to-null (get-in recorder# [:xvfb-recorder :err]))
     (try
       ~@body
       (catch Exception e# (throw e#))
       (finally
         (webdriver.screen/stop-recorder recorder#)
         (webdriver.screen/stop-screen ~screen)))))


src/webdriver/driver_manager.clj
(ns webdriver.driver-manager
  (:gen-class)
  (:require [org.httpkit.client :as http]
            [clojure.java.io :as io]
            [clojure.string :as s]
            [komcrad-utils.io :as kio]
            [me.raynes.fs.compression :refer [gunzip untar]]
            [komcrad-utils.wait :refer [wait-for]]
            [clj-file-zip.core :as zip]
            [me.raynes.conch.low-level :as sh])
  (:import (org.openqa.selenium.logging
            LoggingPreferences LogType)
           (java.util.logging Level Logger)))

(defonce latest-chrome-version "83.0.4103.39")
(defonce latest-gecko-version "0.26.0")

(defn- mkdownload-dir [m]
  (let [dir (:download-dir m)]
    (if dir
      (do (.mkdir (clojure.java.io/file dir))
          (if (not (.isDirectory (clojure.java.io/file dir)))
            (throw (Exception. ":download-dir must be a folder."))
            (.getCanonicalPath (clojure.java.io/file dir))))
      (System/getProperty "user.dir"))))

(defn host-os []
  (let [os (System/getProperty "os.name")]
    (cond
      (s/includes? os "Linux") :linux
      (s/includes? os "Windows") :windows)))

(defn latest-driver-version [m]
  (cond (= :firefox (:driver-type m)) latest-gecko-version
        (= :chrome (:driver-type m)) latest-chrome-version
        :else (latest-driver-version (merge m {:driver-type :chrome}))))

(defn driver-version [m]
  (if (:version m)
    (:version m)
    (latest-driver-version m)))

(defn driver-download-url [m]
  (cond
    (= (:driver-type m) :firefox)
    (str "https://github.com/mozilla/geckodriver/releases/download/v"
         (driver-version m) "/geckodriver-v" (driver-version m)
         "-linux64.tar.gz")
    (= (:driver-type m) :chrome)
    (str "https://chromedriver.storage.googleapis.com/" (driver-version m)
         "/chromedriver_linux64.zip")
    :else (driver-download-url (merge m {:driver-type :chrome}))))

(defn host-os []
  (let [os (System/getProperty "os.name")]
    (cond
      (s/includes? os "Linux") :linux
      (s/includes? os "Windows") :windows)))

(defn driver-download-url [m]
  (cond
    (= (:driver-type m) :firefox)
    (str "https://github.com/mozilla"
         "/geckodriver/releases/download/v"
         (driver-version m) "/geckodriver-v"
         (driver-version m)
         "-linux64.tar.gz")
    (= (:driver-type m) :chrome)
    (str "https://chromedriver.storage.googleapis.com/"
         (driver-version m) "/chromedriver_linux64.zip")
    :else (driver-download-url (merge m {:driver-type :chrome}))))

(defn driver-parent [m]
  (let [base (if (:driver-download-dir m)
               (:driver-download-dir m)
               "./.webdrivers")
        type (cond
               (= (:driver-type m) :firefox) "/geckodriver"
               (= (:driver-type m) :chrome) "/chromedriver"
               :else "/chromedriver")
        arch (cond
               (= (host-os) :linux) "/linux64"
               (= (host-os) :windows) "/win32")
        version (str "/" (driver-version m) "/")]
    (str base type arch version)))

(defn download-driver [m]
  (let [url (driver-download-url m)
        file-name (last (s/split url #"/"))
        down-dir (driver-parent m)
        res (http/get (driver-download-url m))]
    (if (= 404 (:status @res))
      (throw (Exception. (str "Could not find driver " (:driver-type m)
                              " version " (driver-version m))))
      (do
        (io/make-parents (io/file (str down-dir file-name)))
        (io/copy (:body @res) (io/file (str down-dir file-name)))
        (cond
          (= (:driver-type m) :chrome)
          (zip/unzip (str down-dir file-name) down-dir)
          (= (:driver-type m) :firefox)
          (do
            (kio/with-tmps [tar (kio/tmp-file)]
              (gunzip (str down-dir file-name)
                      (.getCanonicalPath tar))
              (untar tar down-dir))))
        (io/delete-file (io/file (str down-dir file-name)))
        (cond
          (= (:driver-type m) :chrome)
          (let [file (io/file (str down-dir) "chromedriver")]
            (.setExecutable file true) file)
          (= (:driver-type m) :firefox)
          (let [file (io/file (str down-dir) "geckodriver")]
            (.setExecutable file true) file))))))

(defn driver-bin [m]
  (let [bin (cond
              (= :firefox (:driver-type m))
              (io/file (str (driver-parent m) "geckodriver"))
              (= :chrome (:driver-type m))
              (io/file (str (driver-parent m) "chromedriver")))]
    (when-not (.exists bin)
      (download-driver m))
    bin))

(defn start-remote-driver [m]
  (let [bin (driver-bin m)
        port (kio/available-port)
        remote-driver
        (if (:xvfb-port m)
          (sh/proc (.getCanonicalPath bin) (str "--port=" port)
                   :env {"DISPLAY" (str ":" (:xvfb-port m))})
          (sh/proc (.getCanonicalPath bin) (str "--port=" port)))]
    (wait-for #(kio/host-port-listening? "localhost" port) 10000 10)
    {:remote-driver remote-driver
     :remote-driver-port port
     :url (str "http://localhost:" port)}))

(defn start-recorded-driver [m]
  (let [bin (driver-bin m)]
    (sh/proc "Xvfb" ":44" "-screen" "0" "1920x1080x24")
    (sh/proc "ffmpeg" "-y" "-f" "x11grab" "-video_size" "1920x1080" "-i" ":44" "-codec:v" "libx264" "-r" "60" "/tmp/glados.mp4")
    (sh/proc (.getCanonicalPath bin) "--port=5555" :env {"DISPLAY" ":44"})))

(defn- chrome-cap [m]
  (let [options (new org.openqa.selenium.chrome.ChromeOptions)
        capabilities (. org.openqa.selenium.remote.DesiredCapabilities chrome)]
    (. options addArguments (concat (if (:driver-args m) (:driver-args m) [])
                                    ["--window-size=1920x1080" "--no-sandbox"]))
    (.setExperimentalOption options "prefs"
                            (doto (new java.util.HashMap)
                              (.put "profile.default_content_settings.popups" 0)
                              (.put "download.default_directory" (mkdownload-dir m))))
    (.setCapability capabilities (. org.openqa.selenium.remote.CapabilityType
                                    ACCEPT_SSL_CERTS) true)
    (.setCapability capabilities (. org.openqa.selenium.remote.CapabilityType
                                    ACCEPT_INSECURE_CERTS) true)
    (.setCapability capabilities (. org.openqa.selenium.chrome.ChromeOptions
                                    CAPABILITY) options)
    capabilities))

(defn create-chrome-driver
  "creates a chromedriver and passes in vector args as command line arguments"
  [m]
  (. System setProperty "wdm.targetPath" ".webdrivers/")
  (. System setProperty "webdriver.chrome.silentOutput", "true")
  (-> (. io.github.bonigarcia.wdm.WebDriverManager chromedriver)
      (.version latest-chrome-version)
      (.setup))
  {:driver (new org.openqa.selenium.chrome.ChromeDriver (chrome-cap m))})

(defn- firefox-cap [m]
  (let [options (new org.openqa.selenium.firefox.FirefoxOptions)
        capabilities (. org.openqa.selenium.remote.DesiredCapabilities firefox)
        profile (new org.openqa.selenium.firefox.FirefoxProfile)]
    (.setPreference profile "browser.download.folderList" 2)
    (.setPreference profile "browser.download.dir" (mkdownload-dir m))
    (.setPreference profile "browser.helperApps.alwaysAsk.force" false)
    (.setPreference profile "browser.helperApps.neverAsk.saveToDisk"
                    (str "text/csv,application/x-msexcel,application/excel,"
                         "application/x-excel,application/vnd.ms-excel,"
                         "image/png,image/jpeg,text/html,text/plain,"
                         "application/msword,application/xml,csv,zip,"
                         "application/x-download,application/zip,"
                         "application/octet-stream"))
    (. options addArguments (if (:driver-args m) (:driver-args m) []))
    (. options setProfile profile)
    (.setCapability
     capabilities
     (. org.openqa.selenium.firefox.FirefoxOptions FIREFOX_OPTIONS) options)
    capabilities))

(defn create-firefox-driver
  "creates a firefoxdriver and passes in vector args as command line arguments"
  [m]
  (. System setProperty "wdm.targetPath" ".webdrivers/")
  (. System setProperty "wdm.geckoDriverVersion" latest-gecko-version)
  (. System setProperty "webdriver.firefox.marionette" "true")
  (. System setProperty "webdriver.firefox.logfile" ".webdrivers/gecko.log")
  (-> (. io.github.bonigarcia.wdm.WebDriverManager firefoxdriver)
      (.version latest-gecko-version)
      (.setup))
  {:driver (new org.openqa.selenium.firefox.FirefoxDriver (firefox-cap m))})

(defn create-remote-driver
  [m]
  (.setLevel (Logger/getLogger "org.openqa.selenium") Level/OFF)
  (let [cap (cond
              (= :chrome (:driver-type m)) (chrome-cap m)
              (= :firefox (:driver-type m)) (firefox-cap m)
              :else (chrome-cap m))]
    {:driver (new org.openqa.selenium.remote.RemoteWebDriver
                  (new java.net.URL (:url m)) cap)}))

(defn set-size [driver width height]
  (.setSize (.window (.manage (:driver driver)))
            (new org.openqa.selenium.Dimension width height)))

(defn headless-remote-driver [m]
  (let [m (merge {:driver-type :chrome} m)
        remote-driver (start-remote-driver m)
        driver (create-remote-driver (dissoc (merge m remote-driver)
                                             :driver-args))]
    (set-size driver 1910 1070)
    (merge driver remote-driver)))
